File: src/features/workouts/restTimer.ts
/**
 * Rest-timer math. Kept pure and time-injectable so it survives app restarts:
 * the store persists an absolute `restEndsAt` epoch and the remaining time is
 * always derived from the current clock, never a decrementing counter.
 */
export function restRemainingSeconds(restEndsAt: number | null, now: number = Date.now()): number {
  if (restEndsAt == null) return 0;
  const ms = restEndsAt - now;
  return ms <= 0 ? 0 : Math.ceil(ms / 1000);
}

export function isResting(restEndsAt: number | null, now: number = Date.now()): boolean {
  return restRemainingSeconds(restEndsAt, now) > 0;
}

/** Format seconds as M:SS. */
export function formatClock(totalSeconds: number): string {
  const s = Math.max(0, Math.floor(totalSeconds));
  const m = Math.floor(s / 60);
  const rem = s % 60;
  return `${m}:${rem.toString().padStart(2, '0')}`;
}

File: android-native/app/src/main/java/com/fitnessrpg/app/domain/workouts/RestTimer.kt
package com.fitnessrpg.app.domain.workouts

import kotlin.math.ceil

/**
 * Rest-timer math. Pure and time-injectable so it survives app restarts: the
 * store persists an absolute [restEndsAt] epoch and the remaining time is always
 * derived from the current clock, never a decrementing counter.
 */
fun restRemainingSeconds(restEndsAt: Long?, now: Long = System.currentTimeMillis()): Int {
    if (restEndsAt == null) return 0
    val ms = restEndsAt - now
    return if (ms <= 0) 0 else ceil(ms / 1000.0).toInt()
}

fun isResting(restEndsAt: Long?, now: Long = System.currentTimeMillis()): Boolean =
    restRemainingSeconds(restEndsAt, now) > 0

/** Format seconds as M:SS. */
fun formatClock(totalSeconds: Int): String {
    val s = maxOf(0, totalSeconds)
    val m = s / 60
    val rem = s % 60
    return "$m:${rem.toString().padStart(2, '0')}"
}
